_posts/2016-10-03-what-i-wish-i-knew.md

```diff
@@ -0,0 +1,96 @@
+
+[Sandro’s book “The Software Craftsman”](https://www.goodreads.com/book/show/23215733-the-software-craftsman) describes how a mentor is beneficial to a new software developer at great lengths.
+
+Depending on the company you are working with, you might be provided with a mentor, or someone will come to you and spontaneously offer their help. Sometimes you will need to look for them. If you cannot find them amongst your colleagues, look for them in community events for example.
+
+##Ask questions##
+I did not know what a framework was when I got out of college. Looking up the definition of Wikipedia did not help either, and for several long months I was too afraid to show my ignorance and ask.
+
+Do not repeat my mistake. You are not expected to arrive at your first job knowing everything. Ask, don’t be afraid to show you do not know something.
+
+Most of the time, you will find out that either at least one other person in the team had the same question, or people have slightly different definitions of the term you are asking about. Even if you are the only one asking the question, you will not be the only one learning something in return.
+
+#Get good practices early#
+There good practices will be the foundation upon which you will build your career. Get them as early as possible.
+
+##Leverage existing libraries and frameworks##
+I barely knew how to properly use a library when I got out of college. As I said earlier I did not even know what a framework was.
+
+As juniors, we have a tendency to re-invent the wheel. Our custom-made solutions will usually be of less quality than a library on which hundreds people have collaborated over several years. Try to re-use what is already available.
+
+On the other hand, understand what are the tradeoffs of a given library. Is it still maintained? It is easy to use? Does it require huge changes in your application’s architecture? Make sure that you isolate that library, so that replacing it does not mean making changes all over your code.
+
+##But still understand the fundamentals##
+Libraries are convenient, because they remove complexity. However, take the time to understand a bit of how they work under the hood.
+
+While your web framework will handle HTTP requests for you, you need to get what HTTP is and how to conveys information between two computers. Similarly, take the time to understand the basis of DNS, TLS, and other low-level protocols. If your application uses OAuth or session-based authentication, you need to know it and you need to know what the consequences are.
+
+##Assume the mistake is yours##
+You will face unexplainable bugs. Sometimes even [heisenbugs](https://en.wikipedia.org/wiki/Heisenbug), that disappear when you debug them. In despair, you might be tempted to reject the fault to something else than your code. Always start with the assumption that it is not an issue with the caller, the library, the OS or anything else.
```

Comment:
By caller I meant a potential client or user, but it's misleading so I'll remove it.